smarts/core/local_traffic_provider.py

```diff
@@ -0,0 +1,1228 @@
+        self._in_front_after_cutin_secs = 0
+        self._cutin_hold_secs = 3
+        self._target_cutin_gap = 2.5 * self._min_space_cush
+        self._aggressiveness = float(self._vtype.get("lcAssertive", 1.0))
+        if self._aggressiveness <= 0:
+            self._logger.warning(
+                "non-positive value {self._aggressiveness} for 'assertive' lane-changing parameter will be ignored"
+            )
+            self._aggressiveness = 1.0
+        self._cutin_prob = float(self._vtype.get("lcCutinProb", 0.0))
+        if not 0.0 <= self._cutin_prob <= 1.0:
+            self._logger.warning(
+                "illegal probability {self._cutin_prob} for 'cutin_prob' lane-changing parameter will be ignored"
+            )
+            self._cutin_prob = 0.0
+        self._dogmatic = bool(self._vtype.get("lcDogmatic", False))
+
+        self._max_angular_velocity = 26
+        self._prev_angular_err = None
+
+        self._owner._actors_created += 1
+
+    @classmethod
+    def from_flow(cls, flow: Dict[str, Any], owner: LocalTrafficProvider):
+        """Factory to construct a _TrafficActor object from a flow dictionary."""
+        vclass = flow["vtype"]["vClass"]
+        dimensions = VEHICLE_CONFIGS[vclass].dimensions
+        vehicle_type = vclass if vclass != "passenger" else "car"
+
+        new_actor = cls(flow, owner)
+        new_actor._lane, new_actor._offset = new_actor._resolve_flow_pos(
+            flow, "depart", dimensions
+        )
+        position = new_actor._lane.from_lane_coord(RefLinePoint(s=new_actor._offset))
+        heading = vec_to_radians(
+            new_actor._lane.vector_at_offset(new_actor._offset)[:2]
+        )
+        init_speed = new_actor._resolve_flow_speed(flow)
+        new_actor._state = VehicleState(
+            vehicle_id=f"{new_actor._vtype['id']}-{new_actor._owner._actors_created}",
+            pose=Pose.from_center(position, Heading(heading)),
+            dimensions=dimensions,
+            vehicle_type=vehicle_type,
+            vehicle_config_type=vclass,
+            speed=init_speed,
+            linear_acceleration=np.array((0.0, 0.0, 0.0)),
+            source=new_actor._owner.source_str,
+            role=ActorRole.Social,
+        )
+        new_actor._dest_lane, new_actor._dest_offset = new_actor._resolve_flow_pos(
+            flow, "arrival", dimensions
+        )
+        return new_actor
+
+    @classmethod
+    def from_state(
+        cls,
+        state: VehicleState,
+        owner: LocalTrafficProvider,
+        route: Optional[RoadMap.Route],
+    ):
+        """Factory to construct a _TrafficActor object from an existing VehiclState object."""
+        cur_lane = owner.road_map.nearest_lane(state.pose.point)
+        if not route or not route.roads:
+            route = owner.road_map.random_route(starting_road=cur_lane.road)
+        route_roads = [road.road_id for road in route.roads]
+        route_id = owner._cache_route_lengths(route_roads)
```

Comment:
OK, I see your point and will rename it.   But just to clarify what I was thinking originally:  `_cache_route_lengths()` returns the "key into the cache" for the route that it just cached.

Re your suggestions, I don't really like using `_generate_route()` there, since route _generation_ only happens _sometimes_ (we'd prefer to use the existing route that's passed in as an argument).   So I may try to come up with alternative naming, but I'll try to honor the spirit of your point to make what's going on clearer.

Reply:
OK, I've changed the name of `route_id` to `route_key` and added a comment about it.  Please let me know if you don't think this is enough.